lib/graphql/query/serial_execution/field_resolution.rb

```diff
@@ -32,7 +32,10 @@ def result
         # continue by "finishing" the value, eg. executing sub-fields or coercing values
         def get_finished_value(raw_value)
           raise raw_value if raw_value.instance_of?(GraphQL::ExecutionError)
-          return nil if raw_value.nil?
+          if raw_value.nil?
+            raise GraphQL::ExecutionError, "Type mismatch resolving '#{field.name}'" if field.type.kind.non_null?
```

Comment:
Right, it looks like GraphQL::Query::BaseExecution::ValueResolution::NonNullResolution doesn't call it at all.  So we could either do this check there, or have that class delegate to NonNullType#coerce_result